x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java

```diff
@@ -0,0 +1,120 @@
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.indices.ShardLimitValidator;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.StreamSupport;
+
+/**
+ * This class upgrades frozen indices to apply the index.shard_limit.group=frozen setting after all nodes have been upgraded to 7.13+
+ */
+public class SearchableSnapshotIndexMetadataUpgrader {
+    private static final Logger logger = LogManager.getLogger(SearchableSnapshotIndexMetadataUpgrader.class);
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final AtomicBoolean upgraded = new AtomicBoolean();
+
+    public SearchableSnapshotIndexMetadataUpgrader(ClusterService clusterService, ThreadPool threadPool) {
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+    }
+
+    public void initialize() {
+        clusterService.addListener(this::clusterChanged);
+    }
+
+    private void clusterChanged(ClusterChangedEvent event) {
+        if (upgraded.get()) {
+            return;
+        }
+
+        if (event.localNodeMaster() && event.state().nodes().getMinNodeVersion().onOrAfter(Version.V_7_13_0)) {
+            // only want one doing this at a time, assume it succeeds and reset if not.
+            if (upgraded.compareAndSet(false, true)) {
+                final Executor executor = threadPool.generic();
+                executor.execute(() -> maybeUpgradeIndices(event.state()));
+            }
+        }
+    }
+
+    private void maybeUpgradeIndices(ClusterState state) {
+        // 99% of the time, this will be a noop, so precheck that before adding a cluster state update.
+        if (needsUpgrade(state)) {
+            logger.info("Upgrading partial searchable snapshots to use frozen shard limit group");
+            clusterService.submitStateUpdateTask("searchable-snapshot-index-upgrader", new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    return upgradeIndices(currentState);
+                }
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    logger.warn("upgrading frozen indices to have frozen shard limit group failed", e);
```

Comment:
maybe add `, will retry on the next cluster state update`?

Reply:
That is, we are letting the user know that something happened, at the warn level. Then they’re going to wonder what do they do? So let’s let them know, you don’t need to take any action, we’ll do it for you. Then, this is a purely informational error message from their perspective.

Reply:
++, fixed in 01c1b3e2e93bebfb2468496f567cbb09be290be5